[RFC] Support complex dtype in random.normal

PhilipVinc opened the pull request:
As I was asking around in #4680 but had no answer, I decided to take a stab at this.

This is a relatively straightforward PR to support complex dtypes in `random.normal`. The rationale of this is that now stax and flax will support out of the box complex-valued networks. This is especially important for flax because it does some peculiar things behind the scenes...

I would also like to do the same for `jax.random.uniform`, but that requires a bit more care to handle default arguments, so I would like to know first that you would accept this PR.

As for the rationale of supporting complex numbers in complex-valued neural networks, in Quantum Mechanics we use this extensively (read: all the time). Moreover, complex numbers arise naturally whenever one tries to model oscillating systems, which is another potential use case of jax together with physically-inspired neural networks. 

Now:
```python
>>> jax.random.normal(jax.random.PRNGKey(0), (2,2), dtype=jax.numpy.complex64)
DeviceArray([[ 0.1389316 +1.1378783j ,  1.3706678 -0.1433144j ],
             [-0.5311612 -0.59153646j,  0.02033774+0.79466224j]],            dtype=complex64)

```

Before:
```python
>>> jax.random.normal(jax.random.PRNGKey(0), (2,2), dtype=jax.numpy.complex64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/filippovicentini/Documents/pythonenvs/netket_env/lib64/python3.8/site-packages/jax/random.py", line 635, in normal
    raise ValueError(f"dtype argument to `normal` must be a float dtype, "
ValueError: dtype argument to `normal` must be a float dtype, got <class 'jax._src.numpy.lax_numpy.complex64'>
```

PhilipVinc commented:
@mattjj Sorry for the bump, but could you tell me if this is something that you'd consider?

VolodyaCO commented:
This is super helpful!

dambuck commented:
Agreed! That would definitely be a helpful feature to have

hawkinsp commented on jax/random.py:
```diff
@@ -631,7 +663,7 @@ def normal(key: jnp.ndarray,
     A random array with the specified shape and dtype.
   """
-  if not dtypes.issubdtype(dtype, np.floating):
+  if not (dtypes.issubdtype(dtype, np.floating) or dtypes.issubdtype(dtype, np.complexfloating)):
```
`np.inexact` covers both floating and complexfloating.

hawkinsp commented on tests/random_test.py:
```diff
@@ -177,6 +178,22 @@ def testRngUniform(self, dtype):
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_dtype={}".format(np.dtype(dtype).name), "dtype": dtype}
+      for dtype in complex_dtypes))
+  def testRngUniform(self, dtype):
```
The duplicate function name appears to be making flake8 sad; I suggest just adding `Complex` to the name. Same below.

hawkinsp commented on jax/random.py:
```diff
@@ -640,6 +672,17 @@ def normal(key: jnp.ndarray,
+    dtype = dtypes.dtype_real(dtype)
+    _re = _normal_real(key_re, shape, dtype)
+    _im = _normal_real(key_im, shape, dtype)
+    return 1/jnp.sqrt(jnp.array(2, dtype=dtype))*(_re + 1j*_im)
```
Nit: The rest of this code base uses spaces around binary operators. Do the same for consistency. Same everywhere.

hawkinsp commented on jax/random.py:
```diff
@@ -374,20 +374,52 @@ def uniform(key: jnp.ndarray,
   shape = abstract_arrays.canonicalize_shape(shape)
   return _uniform(key, shape, dtype, minval, maxval)  # type: ignore
 
+def _uniform_minval_default(dtype: np.dtype) -> [float, complex, jnp.ndarray]:
```
This one feels redundant, in the sense that you could have just written:
`np.array(0, dtype=dtype)`

hawkinsp commented on jax/random.py:
```diff
@@ -374,20 +374,52 @@ def uniform(key: jnp.ndarray,
+  if maxval is None:
+    maxval = _uniform_maxval_default(dtype)
+
+  if not (jnp.issubdtype(dtype, np.floating) or jnp.issubdtype(dtype, np.complexfloating)):
```
np.inexact covers both floating and complexfloating.

See: https://numpy.org/doc/stable/reference/arrays.scalars.html

hawkinsp reviewed (commented):
This seems fine at a high level.

I'm confident that the complex normal support is a good change (since a "complex normal distribution' is a well-understood concept.

I'm less confident that the complex uniform support makes sense as stated. For example, if I simply look what Wikipedia says for "complex uniform", I see: https://en.wikipedia.org/wiki/Complex_random_variable#Uniform_distribution
which is *not* the distribution you have defined here! Is there a reason to think this is the right definition?

PhilipVinc committed:
add complex support to random_normal

PhilipVinc committed:
add tests for complex normal distribution

PhilipVinc committed:
fix flake on ComplexNormal tests

PhilipVinc commented:
@hawkinsp Sorry for the hiatus.

I have removed the commits involving the uniform distribution, and only left the normal, where we all agree on what the implementation should be.

I have also addressed all your comments above, so this is ready for review.

I'll soon open an issue to discuss the API and possible implementation of complex uniform and truncated normal, which are needed to support complex numbers in the initialisers used by stax and flax.
